flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/extraction/TypeInferenceExtractorTest.java

```diff
@@ -0,0 +1,679 @@
+		return Arrays.asList(
+			// function hint defines everything
+			TestSpec
+				.forScalarFunction(FullFunctionHint.class)
+				.expectTypedArguments(DataTypes.INT(), DataTypes.STRING())
+				.expectNamedArguments("i", "s")
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(
+						new String[] {"i", "s"},
+						new ArgumentTypeStrategy[] {
+							InputTypeStrategies.explicit(DataTypes.INT()),
+							InputTypeStrategies.explicit(DataTypes.STRING())}
+					),
+					TypeStrategies.explicit(DataTypes.BOOLEAN())),
+
+			// function hint defines everything with overloading
+			TestSpec
+				.forScalarFunction(FullFunctionHints.class)
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.INT())),
+					TypeStrategies.explicit(DataTypes.INT()))
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.BIGINT())),
+					TypeStrategies.explicit(DataTypes.BIGINT())),
+
+			// global output hint with local input overloading
+			TestSpec
+				.forScalarFunction(GlobalOutputFunctionHint.class)
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.INT())),
+					TypeStrategies.explicit(DataTypes.INT()))
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.STRING())),
+					TypeStrategies.explicit(DataTypes.INT())),
+
+			// unsupported output overloading
+			TestSpec
+				.forScalarFunction(InvalidSingleOutputFunctionHint.class)
+				.expectErrorMessage("Function hints that lead to ambiguous results are not allowed."),
+
+			// global and local overloading
+			TestSpec
+				.forScalarFunction(SplitFullFunctionHints.class)
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.INT())),
+					TypeStrategies.explicit(DataTypes.INT()))
+				.expectOutputMapping(
+					InputTypeStrategies.sequence(InputTypeStrategies.explicit(DataTypes.BIGINT())),
+					TypeStrategies.explicit(DataTypes.BIGINT())),
+
+			// global and local overloading with unsupported output overloading
+			TestSpec
+				.forScalarFunction(InvalidFullOutputFunctionHint.class)
+				.expectErrorMessage("Function hints with same input definition but different result types are not allowed."),
+
+			// invalid data type hint
+			TestSpec
+				.forScalarFunction(IncompleteFunctionHint.class)
+				.expectErrorMessage("Data type hint does neither specify a data type nor input group for use as function argument."),
+
+			// varargs and ANY input group
+			TestSpec
+				.forScalarFunction(ComplexFunctionHint.class)
+				.expectOutputMapping(
+					InputTypeStrategies.varyingSequence(
+						new String[]{"myInt", "myAny"},
+						new ArgumentTypeStrategy[]{InputTypeStrategies.explicit(DataTypes.INT()), InputTypeStrategies.ANY}),
+					TypeStrategies.explicit(DataTypes.BOOLEAN())),
+
+			// ignore argument names during overloading
+			TestSpec
```

Comment:
How is this test different from `InvalidFullOutputFunctionHint`? Do the names make any difference?

If it makes a significant difference can we at least group them/put them one after the other?